### DISCUSSION TITLE
How can I modify the referenced project DefineConstains before everything? Or any way to change the condition for referenced project?

### DISCUSSION THREAD

**Discussion by Flithor** [dotnet/msbuild#9483] [Q&A]
I have a C# project that needs to modify the references (including project references and Nuget packages) of the referenced project itself based on some conditions. I tried to use DefineConstains to do this, but after many attempts, I found it may impossible. The referenced project will always complete all reference-related operations before DefineConstains is modified by main project.  
Maybe I'm on the wrong way... Is there any solution to implement it?

This is the example reference tree about the solution:
(not only one project D, there are many, and some of "Project D" has referenced by Project A)
>Project A (Custom mark "ExcludeC=true")
└ Project B (#if !ExcludeC /\*code\*/ #endif  remove code used NugetPackage C)
　~~└ NugetPackage C (Condition by "ExcludeC", if true then exclude this package)~~
Project D
└ Project B
　└ NugetPackage C (No custom mark "ExcludeC", include it)

**Comment by KalleOlaviNiemitalo** [comment]
Might be doable by adding another project.

Originally:

- A has a ProjectReference on B
- B has a PackageReference on C
- When referenced from A, you want B to have a PackageReference on D instead of C

Change it so that

- A has a ProjectReference on B2
- B2 imports B
- B has a PackageReference on C with Condition="$(MSBuildProjectName) == 'B'"
- B has a PackageReference on D with Condition="$(MSBuildProjectName) == 'B2'"
- B and B2 are in the same directory, so that relative item specs in B refer to the same files even when imported by B2
- B2 sets BaseIntermediateOutPath so that B2 does not use the same obj directory as B

This way NuGet should be able to restore both B and B2 without overwriting files from the other project.

**Reply by Flithor** [reply]
By your means, it should be make a different project to do this. But it might not fit for me, because that I'm try to conditional the reference is that the B and C has been referenced by many other project. Your solution will make the reference tree get mess.

And you may misunderstand my question, this is the sample project reference tree of my project:
>Project A (Custom mark "ExcludeC=true")
└ Project B (#if !ExcludeC /\*code\*/ #endif  remove code used NugetPackage C)
　~~└ NugetPackage C (Condition by "ExcludeC", if true then exclude this package)~~
Project D
└ Project B
　└ NugetPackage C (No custom mark "ExcludeC", include it)

**Reply by KalleOlaviNiemitalo** [reply]
Are these C# projects?

As I see it, the main problem is the obj directory of project B.  NuGet restore writes files there and the content of these files depends on PackageReference items.  NuGet supports having different PackageReference items for each TargetFramework but not other kinds of conditions.  If you need to build B twice for the same TargetFramework but with different PackageReference items, then these builds must have separate obj directories, unless you rerun package restore between the builds.

**Reply by Flithor** [reply]
Yes, are c# project.

But if I set `<DefineConstants>$(DefineConstants);ExcludeC</DefineConstants>` in Project B.csproj, it can exclude Package C when compile.

The only problem is that I didn't find out how to tell Project B by the csproj file of Project A that Package C needs to be excluded before compiling Project B. And not affect the project D.

Now we have to commented `<DefineConstants>$(DefineConstants);ExcludeC</DefineConstants>`  line before compile project D and uncommented this line before compile project A. That's trouble....

`DefineConstants` work well on the preprocessor directives in .cs file, and can also work on `ProjectReference`, but why not work on `PackageRreference`?

**Reply by KalleOlaviNiemitalo** [reply]
<https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#adding-a-packagereference-condition> says only `TargetFramework` is supported in `Condition` of `PackageReference`.

<https://github.com/NuGet/Home/issues/5895> is a feature request for supporting other kinds of conditions but it doesn't look like it's going to be implemented.

**Reply by Flithor** [reply]
But this it works:
```
<PropertyGroup>
    <DefineConstants>$(DefineConstants);ExcludeC</DefineConstants>
</PropertyGroup>

<ItemGroup>
    <PackageReference Condition="!$(DefineConstants.Contains('ExcludeC'))" Include="NugetPackage_C" Version="1.0.0" />
</ItemGroup>
```
`NugetPackage_C` will be exclude from this project and `#if !ExcludeC /**/ #endif` preprocessor directives can exclude the inside code.

Repeat:   
The **only problem** is that I didn't find out a way to let the csproj of Project A tell Project B that needs to exclude Package C before compiling Project B. And not affect the project D which not mark ExcludeC.

Package reference on project B always before reset the DefineConstants on project B.

**Reply by KalleOlaviNiemitalo** [reply]
When NuGet generates a `project.assets.json` file for project B, how do you expect this file to indicate that the reference to `NugetPackage_C` is conditional on DefineConstants?  AFAICT, the file format has no such feature.  It doesn't even list conditions on TargetFramework; instead it lists all target frameworks and then the packages for each framework.

**Reply by Flithor** [reply]
Try this by your self. The condition is works.
```
<PropertyGroup>
    <DefineConstants>$(DefineConstants);ExcludeC</DefineConstants>
</PropertyGroup>

<ItemGroup>
    <PackageReference Condition="!$(DefineConstants.Contains('ExcludeC'))" Include="NugetPackage_C" Version="1.0.0" />
</ItemGroup>
```
Now I need a way to make it work by project A referenced B.

**Reply by KalleOlaviNiemitalo** [reply]
Sure, if DefineConstants is defined in the same project in which the Condition checks it, then NuGet restore will see the value, evaluate the condition, and generate the "assets", "targets", and "props" files accordingly.  But if you want to build variants of the same project and have different PackageReference items in each, then you must do one of these:

- build each variant for a different TargetFramework so that NuGet itself understands how to keep the package lists separate
- use a separate obj directory for each variant so that restoring and building the project for one variant will not interfere with other variants
- restore the packages for one variant, build that, restore the packages for the second variant, build that, etc.; and these MSBuild invocations must be isolated from each other to ensure that nothing is cached in memory from one variant to another.

**Reply by Flithor** [reply]
If there is a way can set all project DefineConstants before everything will be very nice. But looks there is no way...

**Reply by Flithor** [reply]
I've found a workaround is that check the `SolutionName` in condition, it can be work for all project.

**Reply by GangWang01** [reply]
It sounds like project A and project D are not included in the same solution `*.sln` though they are listed in the same tree described in your question. There are separate solutions including the projects that rely on custom mark `ExcludeC=true` and the projects that don't respectively. Is it right?

**Reply by Flithor** [reply]
Project A is referenced by solution X and  project D referenced by solution Y, But solution X need exclude package C.
Conditional by solution name is a final workaround for me, because of this workaround will lead to a new "Things to note" in our work management.
If that's a way to condition by project A will be fine.

**Reply by GangWang01** [reply]
[Global property](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-properties?view=vs-2022#global-properties) set from the command line may help to avoid using solution name.
It looks like no way to condition by project A.
